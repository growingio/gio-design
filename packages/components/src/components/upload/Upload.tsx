import React, { useState, useRef } from 'react';
import RcUpload from 'rc-upload';
import classnames from 'classnames';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';
import {
  IUploadProps,
  ITriggerProps,
  ITriggerMap,
  IProgress,
  IRcFile,
  IUploadFile,
  STATUS_UPLOADING,
  STATUS_SUCCESS,
  STATUS_ERROR,
  TInputUploadType,
} from './interface';
import Toast from '../toast';
import ButtonTrigger from './triggers/ButtonTrigger';
import CardTrigger from './triggers/CardTrigger';
import InputTrigger from './triggers/InputTrigger';
import AvatarTrigger from './triggers/AvatarTrigger';
import DragTrigger from './triggers/DragTrigger';
import { imageFile2DataUrl, fileToObject, getEmptyFileObj, fetchImageFileFromUrl } from './utils';
import xhrRequest from './xhrRequest';
import { UploadPrefixClsContext } from './UploadContext';

const triggerMap: ITriggerMap = {
  button: ButtonTrigger,
  input: InputTrigger,
  card: CardTrigger,
  avatar: AvatarTrigger,
  drag: DragTrigger,
};

const Upload: React.FC<IUploadProps> = ({
  successBorder = false,
  style,
  prefixCls: customPrefixCls,
  className,
  disabled,
  type = 'button',
  inputUploadType = 'url',
  triggerProps,
  action = '.',
  beforeUpload,
  onStart,
  onProgress,
  onError,
  onSuccess,
  onRemove,
  openFileDialogOnClick = true,
  children,
  placeholderImg,
  ...restProps
}: IUploadProps) => {
  const [file, setFile] = useState<IUploadFile>(getEmptyFileObj());

  const rcUploadRef = useRef(null);
  const prefixCls = usePrefixCls('upload', customPrefixCls);

  const rootCls = classnames(className, prefixCls, {
    [`${prefixCls}--disabled`]: disabled,
    [`${prefixCls}--error`]: file?.status === STATUS_ERROR,
    [`${prefixCls}--success`]: file?.status === STATUS_SUCCESS && !successBorder,
    [`${prefixCls}--success-border`]: file?.status === STATUS_SUCCESS && successBorder,
  });

  const Trigger = triggerMap[type];

  const handleBeforeUpload = (fileBeforeUpload: IRcFile, fileList: IRcFile[]) =>
    beforeUpload?.(fileBeforeUpload, fileList);

  const handleStart = (fileOnStart: IRcFile) => {
    const uploadFile: IUploadFile = {
      ...fileToObject(fileOnStart),
      status: STATUS_UPLOADING,
    };
    setFile(uploadFile);
    onStart?.(uploadFile);
  };

  const handleProgress = (step: IProgress, fileOnProgress: IRcFile) => {
    const progressFile: IUploadFile = {
      ...fileToObject(fileOnProgress),
      status: STATUS_UPLOADING,
      percent: step.percent,
    };
    setFile(progressFile);
    onProgress?.(step, fileOnProgress);
  };

  const handleSuccess = async (response: Record<string, unknown>, fileOnSuccess: IRcFile) => {
    let dataUrl = '';
    const uploadFile: IUploadFile = {
      ...fileToObject(fileOnSuccess),
      response,
      status: STATUS_SUCCESS,
    };

    try {
      if (fileOnSuccess.type.startsWith('image/')) {
        dataUrl = await imageFile2DataUrl(fileOnSuccess);
        uploadFile.dataUrl = dataUrl;
      }
    } catch (error) {
      console.log(error);
    }

    setFile(uploadFile);
    onSuccess?.(response, uploadFile);
  };

  const handleError = (error: Error, response: any, fileOnError: IRcFile) => {
    if (type !== 'input') {
      Toast.error('上传失败！');
    }
    const errorFile: IUploadFile = {
      ...fileToObject(fileOnError),
      error,
      response,
      status: STATUS_ERROR,
    };
    setFile(type !== 'drag' ? getEmptyFileObj() : errorFile);
    onError?.(error, errorFile);
  };

  const handleRemove = () => {
    Promise.resolve(typeof onRemove === 'function' ? onRemove(file) : onRemove).then((res) => {
      // 使用者返回了 false，阻止删除操作
      if (res === false) {
        return;
      }

      setFile(getEmptyFileObj());
    });
  };

  const handleInputUpload = async (uploadType: TInputUploadType = 'url', url: string) => {
    const uploadFile = {
      ...file,
      type: 'image',
      name: 'web-image',
    };
    if (uploadType === 'url') {
      uploadFile.dataUrl = url;
      uploadFile.status = STATUS_SUCCESS;
      setFile(uploadFile);
      onSuccess?.({}, uploadFile);
    } else {
      try {
        uploadFile.status = STATUS_UPLOADING;
        setFile(uploadFile);
        const { originFile } = await fetchImageFileFromUrl(url);
        const request = restProps.customRequest || xhrRequest;

        let ac: string;
        if (typeof action === 'function') {
          ac = await action(originFile);
        } else {
          ac = action as string;
        }

        let data: Record<string, string | Blob>;
        if (typeof restProps.data === 'function') {
          data = restProps.data(uploadFile);
        } else {
          data = restProps.data ?? {};
        }
        request({
          action: ac,
          filename: restProps.name ?? 'file',
          data,
          file: originFile,
          headers: restProps.headers ?? {},
          withCredentials: restProps.withCredentials ?? false,
          method: restProps.method || 'post',
          onProgress: (e: IProgress) => handleProgress(e, originFile),
          onSuccess: (res: Record<string, unknown>) => handleSuccess(res, originFile),
          onError: (err: Error, res: Record<string, unknown>) => handleError(err, res, originFile),
        });
      } catch (error) {
        setFile({ ...file, error });
      }
    }
  };

  const rcUploadProps = {
    ...restProps,
    disabled,
    prefixCls,
    action,
    beforeUpload: handleBeforeUpload,
    onStart: handleStart,
    onProgress: handleProgress,
    onSuccess: handleSuccess,
    onError: handleError,
    openFileDialogOnClick: type === 'input' ? false : openFileDialogOnClick,
  };

  const triggerComponentProps: ITriggerProps = {
    triggerProps: {
      ...triggerProps,
      disabled,
    },
    file,
    accept: restProps.accept,
    inputUploadType,
    setFile,
    onRemove: handleRemove,
    onInputUpload: handleInputUpload,
    placeholderImg,
  };

  return (
    <UploadPrefixClsContext.Provider value={prefixCls}>
      <div className={rootCls} style={style}>
        <RcUpload {...rcUploadProps} ref={rcUploadRef}>
          <Trigger {...triggerComponentProps}>{children}</Trigger>
        </RcUpload>
      </div>
    </UploadPrefixClsContext.Provider>
  );
};

export default Upload;
