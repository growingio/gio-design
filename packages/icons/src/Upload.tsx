import * as React from 'react';

function SvgUpload(props: React.SVGProps<SVGSVGElement>) {
  return (
    <svg
      viewBox="0 0 14 14"
      fillRule="evenodd"
      clipRule="evenodd"
      strokeLinejoin="round"
      strokeMiterlimit={1.414}
      width="1em"
      height="1em"
      {...props}
    >
      <path
        d="M14 12.897a1 1 0 01-1 1H1a1 1 0 01-1-1v-2.5a.5.5 0 111 0v2c0 .133.053.26.146.354a.504.504 0 00.354.146h11c.133 0 .26-.053.354-.146a.504.504 0 00.146-.354v-2c0-.133.053-.26.146-.354a.504.504 0 01.708 0 .5.5 0 01.146.354v2.5zM7.5 9.933a.5.5 0 01-1 0v-7.64a.199.199 0 00-.341-.141l-.927.927a.499.499 0 01-.707 0l-.001-.001a.498.498 0 010-.705L6.646.25a.5.5 0 01.708 0l2.121 2.12a.499.499 0 01-.707.708l-.927-.926a.199.199 0 00-.341.14v7.64z"
        fill="currentColor"
      />
    </svg>
  );
}

const MemoSvgUpload = React.memo(SvgUpload);
export default MemoSvgUpload;
