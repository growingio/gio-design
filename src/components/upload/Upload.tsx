import React, { useState, useRef, useEffect } from 'react';
import RcUpload from 'rc-upload';
import classnames from 'classnames';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';
import {
  IUploadProps,
  ITriggerProps,
  ITriggerMap,
  IProgress,
  IRcFile,
  IUploadFile,
  STATUS_UPLOADING,
  STATUS_SUCCESS,
  STATUS_ERROR,
  STATUS_NOT_YET,
  TInputUploadType,
} from './interface';
import ButtonTrigger from './triggers/ButtonTrigger';
import CardTrigger from './triggers/CardTrigger';
import InputTrigger from './triggers/InputTrigger';
import AvatarTrigger from './triggers/AvatarTrigger';
import DragTrigger from './triggers/DragTrigger';
import {
  imageFile2DataUrl,
  fileToObject,
  getEmptyFileObj,
  fetchImageFileFromUrl,
  updateFileList,
  removeFileItem,
  isOnlyAcceptImg,
} from './utils';
import xhrRequest from './xhrRequest';
import { UploadPrefixClsContext } from './UploadContext';
import UploadList from './UploadList';
import Alert from '../alert';

const triggerMap: ITriggerMap = {
  button: ButtonTrigger,
  input: InputTrigger,
  card: CardTrigger,
  avatar: AvatarTrigger,
  drag: DragTrigger,
};

const Upload: React.FC<IUploadProps> = ({
  file: uploadedFile,
  defaultFileList = [],
  successBorder = false,
  style,
  prefixCls: customPrefixCls,
  className,
  disabled,
  type = 'button',
  inputUploadType = 'url',
  triggerProps,
  action = '.',
  beforeUpload,
  onStart,
  onProgress,
  onError,
  onSuccess,
  onRemove,
  openFileDialogOnClick = true,
  children,
  placeholderImg,
  iconSize,
  maxCount = 1,
  directory,
  multiple,
  showUploadList = true,
  ...restProps
}: IUploadProps) => {
  const [file, setFile] = useState<IUploadFile>(getEmptyFileObj(uploadedFile));
  const [showAlert, setAlert] = useState(false);
  // 要上传的文件列表
  const [uploadFileList, setUploadFileList] = useState<IRcFile[]>(defaultFileList.slice(0, maxCount) as IRcFile[]);
  // 已经上传了的文件数量
  const [finish, setFinish] = useState(Math.min(defaultFileList.length, maxCount));
  // 控制dragTrigger是否disabled
  const [uploadDisabled, setUploadDisabled] = useState(disabled);
  useEffect(() => {
    setFile(getEmptyFileObj(uploadedFile));
  }, [uploadedFile]);

  useEffect(() => {
    if (defaultFileList.length >= maxCount) {
      setUploadDisabled(true);
    }
  }, [defaultFileList, maxCount]);

  const rcUploadRef = useRef(null);
  const prefixCls = usePrefixCls('upload', customPrefixCls);

  const rootCls = classnames(className, prefixCls, {
    [`${prefixCls}--disabled`]: disabled || uploadDisabled,
    [`${prefixCls}--error`]: file.status === STATUS_ERROR,
    [`${prefixCls}--success`]: file.status === STATUS_SUCCESS && !successBorder,
    [`${prefixCls}--success-border`]: file.status === STATUS_SUCCESS && successBorder,
    [`${prefixCls}-drag-container`]: type === 'drag',
  });
  const Trigger = triggerMap[type];

  const handleSingleBeforeUpload = (fileBeforeUpload: IRcFile, fileListArgs: IRcFile[]) =>
    beforeUpload?.(fileBeforeUpload, fileListArgs);

  const handleMultipleBeforeUpload = (fileBeforeUpload: IRcFile, fileListArgs: IRcFile[]) => {
    const mergeFileList = [...uploadFileList, ...fileListArgs];
    // 如果选择的文件数量超出最大限制，进行截断
    const fileIndex = mergeFileList.findIndex((item) => item.uid === fileBeforeUpload.uid);
    if (fileIndex >= maxCount) {
      return false;
    }

    setUploadFileList(mergeFileList);

    // 选择的文件超出最大上传文件数量限制
    if (mergeFileList.length >= maxCount) {
      setUploadDisabled(true);
      mergeFileList.length > maxCount && setAlert(true);
      const newFileList = mergeFileList.slice(0, maxCount);
      setUploadFileList(newFileList);
      return beforeUpload?.(fileBeforeUpload, newFileList);
    }

    return beforeUpload?.(fileBeforeUpload, fileListArgs);
  };

  const handleStart = (fileOnStart: IRcFile) => {
    const uploadFile: IUploadFile = {
      ...fileToObject(fileOnStart),
      status: STATUS_UPLOADING,
    };
    setFile(uploadFile);
    onStart?.(uploadFile, uploadFileList);
  };

  const handleProgress = (step: IProgress, fileOnProgress: IRcFile) => {
    const progressFile: IUploadFile = {
      ...fileToObject(fileOnProgress),
      status: STATUS_UPLOADING,
      percent: step.percent,
    };

    setFile(progressFile);
    onProgress?.(step, fileOnProgress, uploadFileList);
  };

  const handleSuccess = async (response: Record<string, unknown>, fileOnSuccess: IRcFile) => {
    let dataUrl = '';
    const uploadFile: IUploadFile = {
      ...fileToObject(fileOnSuccess),
      response,
      status: STATUS_SUCCESS,
      percent: 100,
    };

    try {
      if (fileOnSuccess.type?.startsWith('image/')) {
        dataUrl = await imageFile2DataUrl(fileOnSuccess);
        uploadFile.dataUrl = dataUrl;
      }
    } catch (error) {
      console.error(error);
    }

    const updatedFileList = updateFileList(uploadFile, uploadFileList);

    onSuccess?.(response, uploadFile, updatedFileList.slice(0, finish + 1));
    setFinish(finish + 1);
    setFile(uploadFile);
    setUploadFileList(updatedFileList as IRcFile[]);
  };

  const handleError = (error: Error, response: any, fileOnError: IRcFile) => {
    const errorFile: IUploadFile = {
      ...fileToObject(fileOnError),
      error,
      response,
      status: STATUS_ERROR,
      errorMessage: '文件上传失败！',
    };

    const updatedFileList = updateFileList(errorFile, uploadFileList);
    onError?.(error, errorFile, updatedFileList.slice(0, finish + 1));
    setUploadFileList(updatedFileList as IRcFile[]);
    setFinish(finish + 1);
    setFile(type !== 'drag' ? getEmptyFileObj(uploadedFile) : errorFile);
  };

  const handleRemove = (_file: IUploadFile) => {
    Promise.resolve(typeof onRemove === 'function' ? onRemove(_file) : onRemove).then((res) => {
      // 使用者返回了 false，阻止删除操作
      if (res === false) {
        return;
      }

      const removedFileList = removeFileItem(_file, uploadFileList);

      removedFileList.length < maxCount && setAlert(false);
      removedFileList.length < maxCount && setUploadDisabled(false);
      setFinish(removedFileList.length);
      setUploadFileList(removedFileList as IRcFile[]);

      if (file.dataUrl === uploadedFile?.dataUrl) {
        setFile({
          uid: '',
          size: 0,
          name: '本地上传',
          type: '$empty-file',
          status: STATUS_NOT_YET,
          dataUrl: '',
        });
      } else {
        setFile(getEmptyFileObj(uploadedFile));
      }
    });
  };

  const handleInputUpload = async (uploadType: TInputUploadType = 'url', url: string) => {
    const uploadFile = {
      ...file,
      type: 'image',
      name: 'web-image',
    };
    if (uploadType === 'url') {
      uploadFile.dataUrl = url;
      uploadFile.status = STATUS_SUCCESS;
      setFile(uploadFile);
      onSuccess?.({}, uploadFile, uploadFileList);
    } else {
      try {
        uploadFile.status = STATUS_UPLOADING;
        setFile(uploadFile);
        const { originFile } = await fetchImageFileFromUrl(url);
        const request = restProps.customRequest || xhrRequest;
        let ac: string;
        if (typeof action === 'function') {
          ac = await action(originFile);
        } else {
          ac = action as string;
        }

        let data: Record<string, string | Blob>;
        if (typeof restProps.data === 'function') {
          data = restProps.data(uploadFile);
        } else {
          data = restProps.data ?? {};
        }
        request({
          action: ac,
          filename: restProps.name ?? 'file',
          data,
          file: originFile,
          headers: restProps.headers ?? {},
          withCredentials: restProps.withCredentials ?? false,
          method: restProps.method || 'post',
          onProgress: (e: IProgress) => handleProgress(e, originFile),
          onSuccess: (res: Record<string, unknown>) => handleSuccess(res, originFile),
          onError: (err: Error, res: Record<string, unknown>) => handleError(err, res, originFile),
        });
      } catch (error) {
        setFile({ ...file, error });
        onError?.(error, file, uploadFileList);
      }
    }
  };

  const showBeyondAlert = () =>
    showAlert &&
    maxCount && (
      <Alert
        type="error"
        message={
          isOnlyAcceptImg(restProps.accept)
            ? `图片最多上传${maxCount}张，超过将无法上传`
            : `文件最多上传${maxCount}个，超过将无法上传`
        }
        showIcon
        closeable
        onClose={() => setAlert(false)}
      />
    );

  const renderUploadList = () =>
    showUploadList && (directory || multiple) ? (
      <UploadList items={uploadFileList} onRemove={handleRemove} accept={restProps.accept} />
    ) : null;

  const rcUploadProps = {
    ...restProps,
    disabled: disabled || uploadDisabled,
    prefixCls: type === 'drag' ? `${prefixCls}-drag-container` : prefixCls,
    action,
    directory,
    multiple,
    beforeUpload: directory || multiple ? handleMultipleBeforeUpload : handleSingleBeforeUpload,
    onStart: handleStart,
    onProgress: handleProgress,
    onSuccess: handleSuccess,
    onError: handleError,
    openFileDialogOnClick: type === 'input' ? false : openFileDialogOnClick,
  };

  const triggerComponentProps: ITriggerProps = {
    triggerProps: {
      ...triggerProps,
    },
    file,
    items: uploadFileList,
    finishCount: finish,
    accept: restProps.accept,
    inputUploadType,
    setFile,
    onRemove: handleRemove,
    onReSelect: restProps.onReSelect,
    onInputUpload: handleInputUpload,
    placeholderImg,
    iconSize,
    directory,
    multiple,
    maxCount,
    disabled: disabled === true ? disabled : uploadDisabled,
    dragStyle: restProps.dragStyle,
  };

  if (type === 'drag') {
    return (
      <UploadPrefixClsContext.Provider value={prefixCls}>
        <div>
          <div className={rootCls} style={style}>
            {showBeyondAlert()}
            <RcUpload {...rcUploadProps} ref={rcUploadRef}>
              <Trigger {...triggerComponentProps}>{children}</Trigger>
            </RcUpload>
          </div>
          {renderUploadList()}
        </div>
      </UploadPrefixClsContext.Provider>
    );
  }

  return (
    <UploadPrefixClsContext.Provider value={prefixCls}>
      <div className={rootCls} style={style}>
        <RcUpload {...rcUploadProps} ref={rcUploadRef}>
          <Trigger {...triggerComponentProps}>{children}</Trigger>
        </RcUpload>
      </div>
    </UploadPrefixClsContext.Provider>
  );
};

export default Upload;
