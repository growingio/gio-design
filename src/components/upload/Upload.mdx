import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import Upload from './Upload';

# Upload 上传

<Subtitle>文件选择上传和拖拽上传控件。</Subtitle>

## 参数说明

<ArgsTable of={Upload} />

## 代码演示

### 默认-按钮上传

1. 用户点击按钮弹出文件选择框。若上传的是图片，则在按钮区域显示图片的缩略图，及图片名称，名称显示不全时，hover Tooltip 展示全名。若上传的是文件，则跳转页面。
2. 若上传图片失败，出现 toast 提示失败原因。

[Example](/?path=/story/functional-components-upload--default)

### URL 上传

1. 在输入框内输入图片的 URL，回车即可上传图片。
2. 若上传失败，使用 input 的错误提示方式，说明上传失败的原因。

[Example](/?path=/story/functional-components-upload--url-upload)

### logo 上传

1. 用于上传 logo，点击后弹出文件选择框，上传后图片等比缩放，显示缩略图。
2. 图片上传中套用 Loading 组件。若上传图片失败，出现 toast 提示失败原因。

[Example](/?path=/story/functional-components-upload--card-upload)

### 头像上传

1. 用于上传头像，点击后弹出文件选择框，上传后图片等比缩放，显示缩略图。
2. 图片上传中套用 Loading 组件。若上传图片失败，出现 toast 提示失败原因。

[Example](/?path=/story/functional-components-upload--avatar-upload)

### 上传区域

1. 可以将文件或图片等拖拽到制定区域上传，也可以点击上传。若上传的是图片，上传后图片等比缩放填充满父容器。
2. 图片上传中套用 Loading 组件。若上传失败，出现 toast 提示失败原因。
3. 文件上传中套用 Loading 组件。上传成功/上传失败都在界面上显示。

[Example](/?path=/story/functional-components-upload--area-upload)
