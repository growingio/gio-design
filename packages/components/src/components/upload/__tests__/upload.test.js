import React from 'react';
import { mount } from 'enzyme';
import { setup, clear } from '../tests/mock';
import { mountTest, mountSnapshot } from '../tests/mount';
import Upload from '..';

const uploadTypes = ['button', 'input', 'card', 'avatar', 'drag'];

describe('Testing Upload mount', () => {
  uploadTypes.forEach((type) => {
    mountTest(type);
    mountSnapshot(type);
  });
});

describe('Testing Upload actions', () => {
  beforeEach(() => setup());
  afterEach(() => clear());

  // it('return promise in beforeUpload', (done) => {
  //   const wrapper = mount(
  //     <Upload
  //       action="http://upload.com"
  //       beforeUpload={() => new Promise((resolve) => setTimeout(() => resolve('success'), 100))}
  //       onSuccess={(res, file) => {
  //         if (file.status !== 'uploading') {
  //           done();
  //         }
  //       }}
  //       onError={(error, file) => {
  //         if (file.status !== 'uploading') {
  //           done();
  //         }
  //       }}
  //     />
  //   );
  //   wrapper.find('input').simulate('change', {
  //     target: {
  //       files: [{ file: 'foo.png' }],
  //     },
  //   });
  // });
});
