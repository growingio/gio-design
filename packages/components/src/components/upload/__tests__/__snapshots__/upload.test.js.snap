// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Upload mount match snapshot should match snapshot 1`] = `
<div
  className="gio-customized-upload gio-upload gio-upload--disabled"
>
  <span
    className="gio-upload gio-upload-disabled"
    role="button"
  >
    <input
      directory={null}
      multiple={false}
      onChange={[Function]}
      onClick={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
      webkitdirectory={null}
    />
    <button
      className="gio-btn gio-upload__btn gio-btn-large"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <span
        className="gio-icon"
      >
        <svg
          className="gio-icon-svg"
          clipRule="evenodd"
          fill="currentColor"
          fillRule="evenodd"
          height="1rem"
          strokeLinejoin="round"
          strokeMiterlimit={1.414}
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 14 14"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 12.897a1 1 0 01-1 1H1a1 1 0 01-1-1v-2.5a.5.5 0 111 0v2c0 .133.053.26.146.354a.504.504 0 00.354.146h11c.133 0 .26-.053.354-.146a.504.504 0 00.146-.354v-2c0-.133.053-.26.146-.354a.504.504 0 01.708 0 .5.5 0 01.146.354v2.5zM7.5 9.933a.5.5 0 01-1 0v-7.64a.199.199 0 00-.341-.141l-.927.927a.499.499 0 01-.707 0l-.001-.001a.498.498 0 010-.705L6.646.25a.5.5 0 01.708 0l2.121 2.12a.499.499 0 01-.707.708l-.927-.926a.199.199 0 00-.341.14v7.64z"
            fill="currentColor"
          />
        </svg>
      </span>
      <span
        className="gio-upload__btn-label"
        title="本地上传"
      >
        本地上传
      </span>
    </button>
  </span>
</div>
`;

exports[`Testing Upload mount match snapshot should match snapshot 2`] = `
<div
  className="gio-customized-upload gio-upload gio-upload--disabled"
>
  <span
    className="gio-upload gio-upload-disabled"
    role="button"
  >
    <input
      directory={null}
      multiple={false}
      onChange={[Function]}
      onClick={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
      webkitdirectory={null}
    />
    <button
      className="gio-btn gio-upload__btn gio-btn-large"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <span
        className="gio-icon"
      >
        <svg
          className="gio-icon-svg"
          clipRule="evenodd"
          fill="currentColor"
          fillRule="evenodd"
          height="1rem"
          strokeLinejoin="round"
          strokeMiterlimit={1.414}
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 14 14"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 12.897a1 1 0 01-1 1H1a1 1 0 01-1-1v-2.5a.5.5 0 111 0v2c0 .133.053.26.146.354a.504.504 0 00.354.146h11c.133 0 .26-.053.354-.146a.504.504 0 00.146-.354v-2c0-.133.053-.26.146-.354a.504.504 0 01.708 0 .5.5 0 01.146.354v2.5zM7.5 9.933a.5.5 0 01-1 0v-7.64a.199.199 0 00-.341-.141l-.927.927a.499.499 0 01-.707 0l-.001-.001a.498.498 0 010-.705L6.646.25a.5.5 0 01.708 0l2.121 2.12a.499.499 0 01-.707.708l-.927-.926a.199.199 0 00-.341.14v7.64z"
            fill="currentColor"
          />
        </svg>
      </span>
      <span
        className="gio-upload__btn-label"
        title="本地上传"
      >
        本地上传
      </span>
    </button>
  </span>
</div>
`;

exports[`Testing Upload mount match snapshot should match snapshot 3`] = `
<div
  className="gio-customized-upload gio-upload gio-upload--disabled"
>
  <span
    className="gio-upload gio-upload-disabled"
    role="button"
  >
    <input
      directory={null}
      multiple={false}
      onChange={[Function]}
      onClick={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
      webkitdirectory={null}
    />
    <button
      className="gio-btn gio-upload__btn gio-btn-large"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <span
        className="gio-icon"
      >
        <svg
          className="gio-icon-svg"
          clipRule="evenodd"
          fill="currentColor"
          fillRule="evenodd"
          height="1rem"
          strokeLinejoin="round"
          strokeMiterlimit={1.414}
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 14 14"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 12.897a1 1 0 01-1 1H1a1 1 0 01-1-1v-2.5a.5.5 0 111 0v2c0 .133.053.26.146.354a.504.504 0 00.354.146h11c.133 0 .26-.053.354-.146a.504.504 0 00.146-.354v-2c0-.133.053-.26.146-.354a.504.504 0 01.708 0 .5.5 0 01.146.354v2.5zM7.5 9.933a.5.5 0 01-1 0v-7.64a.199.199 0 00-.341-.141l-.927.927a.499.499 0 01-.707 0l-.001-.001a.498.498 0 010-.705L6.646.25a.5.5 0 01.708 0l2.121 2.12a.499.499 0 01-.707.708l-.927-.926a.199.199 0 00-.341.14v7.64z"
            fill="currentColor"
          />
        </svg>
      </span>
      <span
        className="gio-upload__btn-label"
        title="本地上传"
      >
        本地上传
      </span>
    </button>
  </span>
</div>
`;

exports[`Testing Upload mount match snapshot should match snapshot 4`] = `
<div
  className="gio-customized-upload gio-upload gio-upload--disabled"
>
  <span
    className="gio-upload gio-upload-disabled"
    role="button"
  >
    <input
      directory={null}
      multiple={false}
      onChange={[Function]}
      onClick={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
      webkitdirectory={null}
    />
    <button
      className="gio-btn gio-upload__btn gio-btn-large"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <span
        className="gio-icon"
      >
        <svg
          className="gio-icon-svg"
          clipRule="evenodd"
          fill="currentColor"
          fillRule="evenodd"
          height="1rem"
          strokeLinejoin="round"
          strokeMiterlimit={1.414}
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 14 14"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 12.897a1 1 0 01-1 1H1a1 1 0 01-1-1v-2.5a.5.5 0 111 0v2c0 .133.053.26.146.354a.504.504 0 00.354.146h11c.133 0 .26-.053.354-.146a.504.504 0 00.146-.354v-2c0-.133.053-.26.146-.354a.504.504 0 01.708 0 .5.5 0 01.146.354v2.5zM7.5 9.933a.5.5 0 01-1 0v-7.64a.199.199 0 00-.341-.141l-.927.927a.499.499 0 01-.707 0l-.001-.001a.498.498 0 010-.705L6.646.25a.5.5 0 01.708 0l2.121 2.12a.499.499 0 01-.707.708l-.927-.926a.199.199 0 00-.341.14v7.64z"
            fill="currentColor"
          />
        </svg>
      </span>
      <span
        className="gio-upload__btn-label"
        title="本地上传"
      >
        本地上传
      </span>
    </button>
  </span>
</div>
`;

exports[`Testing Upload mount match snapshot should match snapshot 5`] = `
<div
  className="gio-customized-upload gio-upload gio-upload--disabled"
>
  <span
    className="gio-upload gio-upload-disabled"
    role="button"
  >
    <input
      directory={null}
      multiple={false}
      onChange={[Function]}
      onClick={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
      webkitdirectory={null}
    />
    <button
      className="gio-btn gio-upload__btn gio-btn-large"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <span
        className="gio-icon"
      >
        <svg
          className="gio-icon-svg"
          clipRule="evenodd"
          fill="currentColor"
          fillRule="evenodd"
          height="1rem"
          strokeLinejoin="round"
          strokeMiterlimit={1.414}
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 14 14"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 12.897a1 1 0 01-1 1H1a1 1 0 01-1-1v-2.5a.5.5 0 111 0v2c0 .133.053.26.146.354a.504.504 0 00.354.146h11c.133 0 .26-.053.354-.146a.504.504 0 00.146-.354v-2c0-.133.053-.26.146-.354a.504.504 0 01.708 0 .5.5 0 01.146.354v2.5zM7.5 9.933a.5.5 0 01-1 0v-7.64a.199.199 0 00-.341-.141l-.927.927a.499.499 0 01-.707 0l-.001-.001a.498.498 0 010-.705L6.646.25a.5.5 0 01.708 0l2.121 2.12a.499.499 0 01-.707.708l-.927-.926a.199.199 0 00-.341.14v7.64z"
            fill="currentColor"
          />
        </svg>
      </span>
      <span
        className="gio-upload__btn-label"
        title="本地上传"
      >
        本地上传
      </span>
    </button>
  </span>
</div>
`;
